File: Trace2JSON/Trace2JSON/TraceUtility/TraceUtility.m
//
//  TraceUtility.m
//  Trace2JSON
//
//  Created by Itay Brenner on 6/20/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

#import "TraceUtility.h"
#import "InstrumentsPrivateHeaders.h"
#import "ParserFactory.h"
#import "XRDevice+NSDictionary.h"
#import "PFTProcess+NSDicionary.h"

@interface TraceUtility ()
@property (nonatomic, strong) PFTTraceDocument *document;
@end

@implementation TraceUtility

- (void) openDocument:(NSString *)path {
    NSError *error = nil;
    self.document = [[PFTTraceDocument alloc]initWithContentsOfURL:[NSURL fileURLWithPath:path] ofType:@"com.apple.instruments.trace" error:&error];
    if (error) {
        Print(@"Error: %@", error);
        exit(1);
    }
}

- (void) processDocument {
    NSMutableDictionary *resultDictionary = [NSMutableDictionary new];
    
    // This logic was based on Qusic's TraceUtility (https://github.com/Qusic/TraceUtility)
    XRDevice *device = self.document.targetDevice;
    [resultDictionary setObject:[device dictionary] forKey:@"device"];

    PFTProcess *process = self.document.defaultProcess;
    [resultDictionary setObject:[process dictionary] forKey:@"process"];
    
    
    NSMutableDictionary *instruments = [NSMutableDictionary new];
    XRTrace *trace = self.document.trace;
    for (XRInstrument *instrument in trace.allInstrumentsList.allInstruments) {
        NSDictionary *instrumentResult = [self processInstrument:instrument];
        
        if(instrumentResult) {
            [instruments setObject:instrumentResult forKey:instrument.type.uuid];
        }
    }
    
    [resultDictionary setObject:instruments forKey:@"instruments"];
    
    NSError *error = nil;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:resultDictionary options:NSJSONWritingPrettyPrinted error:&error];
    if(!error) {
        Print(@"%@", [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding]);
    }
    
    [self.document close];
}

- (XRContext *) createContext:(XRAnalysisCoreDetailNode *) detailNode container:(XRAnalysisCoreDetailViewController *) detailController {
    if (!detailNode)
        return nil;

    NSString *label = [detailNode label];
    XRAnalysisCoreDetailNode *parentNode = detailNode.parent;
    XRContext *parentContext = [self createContext:parentNode container: detailController];

    XRContext *context = [[XRContext alloc] initWithLabel:label
                                                    value:detailNode
                                               attributes:nil
                                                container:detailController
                                            parentContext:parentContext];

    return context;
}

- (NSDictionary *) processInstrument: (XRInstrument *) instrument {
    // Each instrument can have multiple runs.
    NSArray<XRRun *> *runs = instrument.allRuns;
    if (runs.count == 0) {
        return nil;
    }
    
    NSMutableDictionary *instrumentDictionary = [NSMutableDictionary new];
    [instrumentDictionary setObject:instrument.type.uuid forKey:@"type"];
    [instrumentDictionary setObject:@(runs.count) forKey:@"runsCount"];
    
    NSMutableArray *runsParsed = [NSMutableArray new];
    for (XRRun *run in runs) {
        instrument.currentRun = run;

        // Common routine to obtain contexts for the instrument.
        NSMutableArray<XRContext *> *contexts = [NSMutableArray array];
        if (![instrument isKindOfClass:XRLegacyInstrument.class]) {
            XRAnalysisCoreStandardController *standardController = [[XRAnalysisCoreStandardController alloc]initWithInstrument:instrument document:self.document];
            instrument.viewController = standardController;
            [standardController instrumentDidChangeSwitches];
            [standardController instrumentChangedTableRequirements];
            XRAnalysisCoreDetailViewController *detailController = Ivar(standardController, _detailController);
            [detailController restoreViewState];
            XRAnalysisCoreDetailNode *detailNode = Ivar(detailController, _firstNode);
            while (detailNode) {
                [contexts addObject:[self createContext:detailNode container:detailController]];
                detailNode = detailNode.nextSibling;
            }
        }

        NSString *instrumentID = instrument.type.uuid;
        id<ParserProtocol> parser = [[ParserFactory new] parserForInstrument:instrumentID];
        if (parser) {
            NSDictionary *parserResult = @{
                @"run": @(run.runNumber),
                @"runName": run.displayName,
                @"result": [parser parseContext:contexts withRun:run]
            };
            [runsParsed addObject:parserResult];
        } else {
            NSDictionary *parserResult = @{
                @"run": @(run.runNumber),
                @"runName": run.displayName,
                @"result": @"unsupported instrument",
                @"unsupported": @(YES)
            };
            [runsParsed addObject:parserResult];
            Print(@"Data processor has not been implemented for the instrument: %@", instrument.type.uuid);
        }
    }
    [instrumentDictionary setObject:runsParsed forKey:@"runs"];
    
    if (![instrument isKindOfClass:XRLegacyInstrument.class]) {
        [instrument.viewController instrumentWillBecomeInvalid];
        instrument.viewController = nil;
    }
    
    return instrumentDictionary;
}

@end

File: Trace2JSON/Trace2JSON/TraceUtility/TraceUtility.swift
//
//  TraceUtility.swift
//  Trace2JSON
//
//  Created by Itay Brenner on 6/22/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

import Foundation

class TraceUtility {
    private var document: PFTTraceDocument
    private var parserFactory = ParserFactory()
    private var process: String?
    private var pid: String?
    
    struct TraceConstants {
        static let instrumentsTrace = "com.apple.instruments.trace"
        static let detailController = "_detailController"
        static let firstNode = "_firstNode"
    }
    
    init(path: String, process: String?, pid: String?) throws {
        document = try PFTTraceDocument(contentsOf: URL(fileURLWithPath: path), ofType: TraceConstants.instrumentsTrace)
        self.process = process
        self.pid = pid
    }
    
    func processDocument() -> Trace {
        let trace = Trace()
        
        trace.device = Device(device: document.targetDevice())
        trace.process = SelectedProcess(process: document.defaultProcess())
        
        guard let xrtrace = document.trace(),
            let instruments = xrtrace.allInstrumentsList() else {
            return trace
        }
        
        for instrument in instruments.allInstruments() {
            if let result = processInstrument(instrument) {
                trace.instruments.append(result)
            }
        }
        
        return trace
    }
    
    func processInstrument(_ instrument: XRInstrument) -> Instrument? {
        guard let runs = instrument.allRuns() else {
            return nil
        }
        
        let result = Instrument()
        result.runsCount = runs.count
        result.type = instrument.type()?.uuid()
        
        for run in runs {
            instrument.setCurrentRun(run)
            
            var contexts: [XRContext] = []
            if let _ = instrument as? XRLegacyInstrument {}
            else {
                guard let standardController = XRAnalysisCoreStandardController(instrument: instrument,
                                                                                document: document) else {
                    continue
                }
                instrument.setViewController(standardController)
                standardController.instrumentDidChangeSwitches()
                standardController.instrumentChangedTableRequirements()
                
                guard let detailController: XRAnalysisCoreDetailViewController = RuntimeHacks.getIvar(instance: standardController,
                                                                                                      name: TraceConstants.detailController) else {
                    continue
                }
                detailController.restoreViewState()
                
                guard var detailNode: XRAnalysisCoreDetailNode? = RuntimeHacks.getIvar(instance: detailController,
                                                                                       name: TraceConstants.firstNode) else {
                    continue
                }
                
                while(detailNode != nil) {
                    if let resultContext = createContext(detailNode!, detailController) {
                        contexts.append(resultContext)
                    }
                    detailNode = detailNode?.nextSibling()
                }
            }
            
            if let instrumentId = instrument.type()?.uuid(),
                let parser = parserFactory.parserForInstrument(instrument: instrumentId) {
                result.runs.append(parseContexts(contexts, run: run, parser: parser))
            }
            
            if let _ = instrument as? XRLegacyInstrument {}
            else {
                instrument.viewController()?.instrumentWillBecomeInvalid()
                instrument.setViewController(nil)
            }
        }
        
        
        return result
    }
    
    func createContext(_ detailNode: XRAnalysisCoreDetailNode?, _ detailController: XRAnalysisCoreDetailViewController) -> XRContext? {
        guard let detailNode = detailNode else {
            return nil
        }
        
        return XRContext(label: detailNode.label,
                         value: detailNode,
                         attributes: nil,
                         container: detailController,
                         parentContext: createContext(detailNode.parent, detailController))
    }
    
    func parseContexts(_ contexts: [XRContext], run: XRRun, parser: ParserProtocol) -> InstrumentRun {
        if let process = process {
            return parser.parseContext(contexts, run: run, process: process)
        } else if let pid = pid {
            return parser.parseContext(contexts, run: run, pid: pid)
        } else {
            return parser.parseContext(contexts, run: run)
        }
    }
}
